readme.md

```diff
@@ -1,6 +1,254 @@
+but limits the total autograder runtime for an individual submission to 
+20 minutes,
+at which point the autograder crashes and the submission receives no credit.
+
+This module allows tests to be given time limits, 
+by running tests in their own processes.
+Crucially, stdout and any exceptions are redirected to give feedback to
+students.
+
+
+### fcn-tests	
+
+We found that our tests of students' code tended to involve verifying that a
+long list of inputs gave the correct output.
+Since python's unittest framework lends itself better to more individual 
+test cases, 
+I created essentially a factory for test methods.
+The factory builds these methods primarily 
+from a list of inputs or a dict of inputs mapped to expected outputs.
+
+
+### correct-with-git
+
+Often a student would make a relatively minor mistake that would drastically
+affect their autograder score.
+Moreover, when trying to understand students' code,
+it helped to have the ability to make changes and see how the test results were
+affected -- essentially providing a debugging process for the graders.
+
+This was accomplished by version controlling students' submissions using git,
+and pulling any changes that might have been made by the grader on the remote.
+
+* Create a branch for each student
+* Create a different repo for each assignment
+  * Could also have named branches with assignmentname/studentname
+* Stage and commit all files (and changes thereto) in the submission directory,
+  including metadata
+* Push to the remote
+* Attempt to pull any changes from the remote
+  * This will only be successful if the student has NOT updated their
+	submission (i.e., there was nothing to push).  
+	Otherwise nothing will be pulled.
+  * Any changes that are pulled will be changes the grader made to the
+	assignment
+
+Notes on scoring:
+
+* Obviously correcting students' code will increase their score
+* Typically the grader will want the student to partial (or zero) credit
+  for the mistakes that were corrected
+* In my implementation this was done entirely manually, 
+  in a dedicated manually graded "problem" called something like 
+  "Modifications to the autograder score"
+* It would be possible for these score modifications to be done in some
+  automated way via git (perhaps in a json file or something),
+  but I haven't implemented anything of that sort
+
+What the student sees:
+
+* If the student weren't notified somehow, they wouldn't realize 
+  the code being graded by the autograder was a modified version of 
+  their submission
+* I accomplished this through a printout of the git log 
+  (which displays the commit messages)
+  in a dedicate autograder test
+  * This depends upon the grader to create descriptive commit messages
+    detailing the changes made
+
+
+## Using this package
+
+### Getting started
+
+First, see Gradescope's [autograder documentation][gs-autograder-docs],
+which explains how to use and customize the autograder.
```

Comment:
These lines can be replaced (with a reference to?) or supplemented by the proposed overview of the stages of the autograder (as mentioned in a comment above).  Hopefully this will make for a more useful 'getting started' guide.